src/pg/pg-store.ts

```diff
@@ -0,0 +1,442 @@
+        WHERE smart_contracts.principal = ${args.contractPrincipal}
+          AND tokens.token_number = ${args.tokenNumber}
+      `;
+      if (tokenIdRes.count === 0) {
+        throw new TokenNotFoundError();
+      }
+      if (args.locale && !(await this.isTokenLocaleAvailable(tokenIdRes[0].id, args.locale))) {
+        throw new TokenLocaleNotFoundError();
+      }
+      return await this.getTokenMetadataBundle(tokenIdRes[0].id, args.locale);
+    });
+  }
+
+  /**
+   * Writes a full bundle of token info and metadata (including attributes and properties) into the
+   * db.
+   * @param id - token id
+   * @param values - update bundle values
+   */
+  async updateProcessedTokenWithMetadata(args: {
+    id: number;
+    values: DbProcessedTokenUpdateBundle;
+  }): Promise<void> {
+    await this.sqlWriteTransaction(async sql => {
+      // Update token and clear old metadata (this will cascade into all properties and attributes)
+      await sql`
+        UPDATE tokens SET ${sql(args.values.token)}, updated_at = NOW() WHERE id = ${args.id}
+      `;
+      await sql`DELETE FROM metadata WHERE token_id = ${args.id}`;
+      // Write new metadata
+      if (args.values.metadataLocales && args.values.metadataLocales.length > 0) {
+        for (const locale of args.values.metadataLocales) {
+          const metadataInsert = await sql<{ id: number }[]>`
+            INSERT INTO metadata ${sql(locale.metadata)} RETURNING id
+          `;
+          const metadataId = metadataInsert[0].id;
+          if (locale.attributes && locale.attributes.length > 0) {
+            const values = locale.attributes.map(attribute => ({
+              ...attribute,
+              metadata_id: metadataId,
+            }));
+            await sql`INSERT INTO metadata_attributes ${sql(values)}`;
+          }
+          if (locale.properties && locale.properties.length > 0) {
+            const values = locale.properties.map(property => ({
+              ...property,
+              metadata_id: metadataId,
+            }));
+            await sql`INSERT INTO metadata_properties ${sql(values)}`;
+          }
+        }
+      }
+    });
+  }
+
+  async updateJobStatus(args: { id: number; status: DbJobStatus }): Promise<void> {
+    await this.sql`
+      UPDATE jobs
+      SET status = ${args.status}
+      WHERE id = ${args.id}
+    `;
+  }
+
+  async increaseJobRetryCount(args: { id: number }): Promise<number> {
+    const result = await this.sql<{ retry_count: number }[]>`
+      UPDATE jobs
+      SET retry_count = retry_count + 1
+      WHERE id = ${args.id}
+      RETURNING retry_count
+    `;
+    return result[0].retry_count;
+  }
+
+  /**
+   * Retrieves a number of queued jobs so they can be processed immediately.
+   * @param limit - number of jobs to retrieve
+   * @returns `DbJob[]`
+   */
+  async getPendingJobBatch(args: { limit: number }): Promise<DbJob[]> {
+    return this.sql<DbJob[]>`
+      SELECT * FROM jobs
```

Comment:
Is it a good time to remove the `select *` statements?

Reply:
Done